framework/include/materials/MaterialData.h

```diff
@@ -101,68 +66,95 @@ class MaterialData
+  }
 
   /// Returns true if the AD material property exists - defined by any material.
   template <typename T>
-  bool haveADProperty(const std::string & prop_name) const;
+  bool haveADProperty(const std::string & prop_name) const
+  {
+    return haveGenericProperty<T, true>(prop_name);
+  }
 
+  /**
+   * Retreives a material property
+   * @tparam T The type of the property
+   * @tparam is_ad Whether or not the property is AD
+   * @param prop_name The name of the property
+   * @param state The time state (0 = current, 1 = old, etc; defaults to 0)
+   * @param requestor The MooseObject requesting the property
+   * @return The property for the supplied type and name
+   */
   template <typename T, bool is_ad>
-  bool haveGenericProperty(const std::string & prop_name) const
+  GenericMaterialProperty<T, is_ad> & getProperty(const std::string & prop_name,
+                                                  const unsigned int state,
+                                                  const MooseObject & requestor)
   {
-    if constexpr (is_ad)
-      return haveADProperty<T>(prop_name);
-    else
-      return haveProperty<T>(prop_name);
+    return getPropertyHelper<T, is_ad, false>(prop_name, state, requestor);
   }
-
   /**
-   * @{ Methods for retieving a MaterialProperty object
+   * Retreives a material property
    * @tparam T The type of the property
+   * @tparam is_ad Whether or not the property is AD
    * @param prop_name The name of the property
+   * @param requestor The MooseObject declaring the property
    * @return The property for the supplied type and name
    */
   template <typename T, bool is_ad>
-  auto & getGenericProperty(const std::string & prop_name)
+  GenericMaterialProperty<T, is_ad> & declareProperty(const std::string & prop_name,
+                                                      const MooseObject & requestor)
   {
-    if constexpr (is_ad)
-      return getADProperty<T>(prop_name);
-    else
-      return getProperty<T>(prop_name);
+    return getPropertyHelper<T, is_ad, true>(prop_name, 0, requestor);
   }
-  template <typename T>
-  MaterialProperty<T> & getProperty(const std::string & prop_name);
-  template <typename T>
-  ADMaterialProperty<T> & getADProperty(const std::string & prop_name);
-  template <typename T>
-  MaterialProperty<T> & getPropertyOld(const std::string & prop_name);
-  template <typename T>
-  MaterialProperty<T> & getPropertyOlder(const std::string & prop_name);
-  ///@}
 
   /**
    * Returns true if the stateful material is in a swapped state.
    */
-  bool isSwapped();
+  bool isSwapped() const { return _swapped; }
 
   /**
    * Provide read-only access to the underlying MaterialPropertyStorage object.
    */
   const MaterialPropertyStorage & getMaterialPropertyStorage() const { return _storage; }
 
+  /**
+   * Key that provides access to only the XFEM class.
+   */
+  class XFEMKey
+  {
+    friend class XFEM;
+    XFEMKey() {}
+    XFEMKey(const XFEM &) {}
+  };
```

Comment:
Why are we having to do this? I really don't like seeing names from a downstream library

Reply:
This was when the PR really "clicked" for me; I hadn't been able to figure out how to get to a clean design here without breaking downstream code, but "get a clean design for almost everyone but keep a couple hacks for the exceptions" is at least a step in the right direction IMHO.

Reply:
Should some XFEM stuff be moved upstream though? I know we have `XFEMInterface` and `FEProblemBase::haveXFEM` and various other XFEM references already in the framework.

That should be outside of the scope of this PR though

Reply:
And moving that upstream would not "fix" the hack you need to do here I presume

Reply:
Yeah... I spent a long time trying to clean this up some. It just requires a rewrite. I did _remove_ and restrict access more by removing the raw pointer construction in storage. So... regardless it is an improvement. And I've written nasty things about where this is done